agent/connect/ca/provider_consul.go

```diff
@@ -31,94 +34,109 @@ type ConsulProviderStateDelegate interface {
 
 // NewConsulProvider returns a new instance of the Consul CA provider,
 // bootstrapping its state in the state store necessary
-func NewConsulProvider(rawConfig map[string]interface{}, delegate ConsulProviderStateDelegate) (*ConsulProvider, error) {
-	conf, err := ParseConsulCAConfig(rawConfig)
-	if err != nil {
-		return nil, err
-	}
-	provider := &ConsulProvider{
-		config:   conf,
+func NewConsulProvider(delegate ConsulProviderStateDelegate) *ConsulProvider {
+	return &ConsulProvider{
 		delegate: delegate,
-		id:       fmt.Sprintf("%s,%s", conf.PrivateKey, conf.RootCert),
 	}
+}
+
+func (c *ConsulProvider) Configure(clusterId string, isRoot bool, rawConfig map[string]interface{}) error {
+	// Parse the raw config and update our ID.
+	config, err := ParseConsulCAConfig(rawConfig)
+	if err != nil {
+		return err
+	}
+	c.config = config
+	c.id = fmt.Sprintf("%s,%s", config.PrivateKey, config.RootCert)
 
-	// Check if this configuration of the provider has already been
-	// initialized in the state store.
-	state := delegate.State()
-	_, providerState, err := state.CAProviderState(provider.id)
+	// Exit early if the state store has an entry for this provider's config.
+	_, providerState, err := c.delegate.State().CAProviderState(c.id)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	// Exit early if the state store has already been populated for this config.
 	if providerState != nil {
-		return provider, nil
+		return nil
 	}
 
+	// Write the provider state to the state store.
 	newState := structs.CAConsulProviderState{
-		ID: provider.id,
+		ID:     c.id,
+		IsRoot: isRoot,
```

Comment:
We shouldn't need to store this. One of the comments on RFC was about ensuring that provider implementations (e.g. later external processes) shouldn't need to persistently remember which mode they are running in (root or not) since we always provide it to the Configure call so it can just be kept in memory in the Provider struct.

That said, there are cases where a config might change or even a DC might go from primary to secondary without changing their CA config and we should be able to do the right thing or produce a meaningful error if not.

There is no actual harm in storing this in the internal case, the only argument against really is that we are doing something that other providers won't be able to do which might make us miss edge cases that built-in provider can recover easily from but others can't unless we implicitly require them to store this state.

Do you think it's worth _not_ storing it just to make sure we can handle all the edge cases without? For example we could in Configure include `isRoot` in the (hashed) provider ID so a reconfiguration from primary to secondary even with the same key present would cause a reconfiguration as expected.

Reply:
In fact I think we _need_ isRoot to be part of the id hash. Consider the following:

 * two OSS datacenters setup running independent CAs (e.g. before Multi-DC release) but have same CA key configured via config file
 * upgrade to multi-dc: dc2 is now a secondary but with the same key

In this case DC2's state already has a CA provider state that will match the ID based on the code here even though it's now a secondary and the Consul code will expect it to behave like one - it will says it's initialized and so never get an intermediate signed by the other root. Generally our code will likely hit bugs since it's in secondary mode but operating on CA state that was stored as if it was primary mode!